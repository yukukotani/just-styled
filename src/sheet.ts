import fnv1a from "@sindresorhus/fnv1a";
import type { CSSProperties } from "react";
import { transformTokenToCssVariable } from "./theme";
import { toKebabCase } from "./utils";

type Sheet = {
	className: string;
	css: string;
};

export function generateSheets(props: CSSProperties) {
	const sheets: Sheet[] = [];

	for (const [rawKey, rawValue] of Object.entries(props)) {
		if (rawValue != null && rawValue !== "") {
			const hash = fnv1a(`${rawKey}${rawValue}`, { size: 32 })
				.toString(36)
				.slice(0, 4);
			// appends prefix to avoid a leading number which is invalid as a className
			const className = `j${hash}`;

			const key = toKebabCase(rawKey);
			const value = transformValue(rawValue);
			sheets.push({
				className: className,
				css: `.${className} { ${key}: ${value}; }`,
			});
		}
	}

	return sheets;
}

function transformValue(value: unknown) {
	if (typeof value === "string" && value.startsWith("$")) {
		const cssVariableName = transformTokenToCssVariable(value.slice(1));
		return `var(${cssVariableName})`;
	}

	return String(value);
}
