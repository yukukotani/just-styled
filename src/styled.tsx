import { ComponentProps, CSSProperties, ElementType } from "react";
import fnv1a from "@sindresorhus/fnv1a";

export function styled<C extends ElementType<{ className?: string }>>(
  Component: C,
  props: CSSProperties
) {
  const sheets = generateSheets(props);

  return (props: ComponentProps<C>) => {
    return (
      <>
        {sheets.map((s) => (
          <style key={s.className} href={s.className} precedence="medium">
            {s.sheet}
          </style>
        ))}
        {/* @ts-expect-error */}
        <Component
          {...props}
          className={sheets.map((s) => s.className).join(" ")}
        />
      </>
    );
  };
}

function generateSheets(props: CSSProperties) {
  const sheets: { className: string; sheet: string }[] = [];

  for (const [key, value] of Object.entries(props)) {
    if (value) {
      const kebab = toKebabCase(key);
      const hash = fnv1a(`${key}${value}`, { size: 32 })
        .toString(36)
        .slice(0, 4);
      const className = `x${hash}`;
      sheets.push({
        className: className,
        sheet: `.${className} { ${kebab}: ${value}; }`,
      });
    }
  }

  return sheets;
}
function toKebabCase(str: string) {
  return str.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
}
