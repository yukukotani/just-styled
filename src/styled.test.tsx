import { render, waitFor } from "@testing-library/react";
import React, { type ComponentProps } from "react";
import { beforeEach, describe, expect, test, vi } from "vitest";
import { generateSheets } from "./sheet";
import { styled } from "./styled";

describe("styled", async () => {
  beforeEach(() => {
    vi.clearAllMocks();
    vi.mock("./sheet", { spy: true });
  });

  test("renders style tag in head", async () => {
    const RedDiv = styled("div", {
      color: "red",
      gap: "$spaces.md",
    });

    render(<RedDiv>RedDiv</RedDiv>);

    await waitFor(() => {
      expect(generateSheets).toHaveBeenCalledTimes(1);
      expect(generateSheets).toHaveBeenCalledWith({
        color: "red",
        gap: "$spaces.md",
      });

      const styleTags = Array.from(
        document.getElementsByTagName("style"),
      ).filter((e) => e.dataset.href != null);
      expect(styleTags).toHaveLength(2);
      for (const styleTag of styleTags) {
        expect(document.head).toContainElement(styleTag);
        expect(styleTag).toHaveAttribute("data-precedence", "medium");
        expect(styleTag.textContent).toMatchSnapshot();
      }
    });
  });

  test("renders dynamic style", async () => {
    const YellowDiv = styled("div", {
      color: "yellow",
    });

    render(
      <YellowDiv style={{ backgroundColor: "gray", padding: "$spaces.md" }}>
        RedDiv
      </YellowDiv>,
    );

    await waitFor(() => {
      expect(generateSheets).toHaveBeenCalledTimes(2);
      expect(generateSheets).toHaveBeenCalledWith({ color: "yellow" });
      expect(generateSheets).toHaveBeenCalledWith({
        backgroundColor: "gray",
        padding: "$spaces.md",
      });
    });
  });

  test("renders styled custom component", async () => {
    const BlueComp = styled(Comp, {
      color: "blue",
    });

    const res = render(
      <BlueComp data-testid="blue-comp" title="Blue">
        BlueComp
      </BlueComp>,
    );
    const el = await res.findByTestId("blue-comp");

    await waitFor(() => {
      expect(generateSheets).toHaveBeenCalledTimes(1);
      expect(generateSheets).toHaveBeenCalledWith({ color: "blue" });

      expect(el.className).not.toBe("");
      expect(el).toHaveAttribute("data-title", "Blue");
    });
  });
});

function Comp({
  children,
  title,
  ...props
}: ComponentProps<"div"> & { title: string }) {
  return (
    <div {...props} data-title={title}>
      <span>{title}</span>
      <div>{children}</div>
    </div>
  );
}
