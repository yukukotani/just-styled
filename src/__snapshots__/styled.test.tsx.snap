// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`styled > renders style tag in head 1`] = `".j1dqz { color: red; }"`;

exports[`styled > renders style tag in head 2`] = `".jyt81 { gap: var(--j-spaces-md); }"`;
